Add doc page about joint kinematics (#1497)
ros-controls/ros2_control: hardware_interface/doc/joints_userdoc.rst

Kinematic Loops
  On the other hand, kinematic loops, also known as closed-loop mechanisms, involve several joints that are connected in a kinematic chain and being actuated together. This means that the joints are coupled and cannot be moved independently: In general, the number of DOFs is smaller than the number of joints.
  This structure is typical for parallel kinematic mechanisms, where the end-effector is connected to the base by several kinematic chains.

  An example is the four-bar linkage, which consists of four links and four joints. It can have one or two actuators and consequently one or two DOFs, despite having four joints. Furthermore, we can say that we have one (two) actuated joint and three (two) passive joints, which must satisfy the kinematic constraints of the mechanism.

URDF
#############

URDF is the default format to describe robot kinematics in ROS. However, only serial kinematic chains are supported, except for the so-called mimic joints.
See the `URDF specification <http://wiki.ros.org/urdf/XML/joint>`__ for more details. For the mimic joints, we can read

>   This tag is used to specify that the defined joint mimics another existing joint. The value of this joint can be computed as value = multiplier * other_joint_value + offset.

Author: I copied this over as a quote, should we specify this here?

>   This tag is used to specify that the defined joint mimics another existing joint. The value of this joint can be computed as value = multiplier * other_joint_value + offset.

Reviewer: ```suggestion
  This tag is used to specify that the defined joint mimics another existing joint in terms of position and velocity.
```
Either we mention is specifically or we remove it

>   This tag is used to specify that the defined joint mimics another existing joint. The value of this joint can be computed as value = multiplier * other_joint_value + offset.

Reviewer: Well as it is explained in the other parts of the text, may be we can remove it?

>   This tag is used to specify that the defined joint mimics another existing joint. The value of this joint can be computed as value = multiplier * other_joint_value + offset.

Author: what do you think about the last update?

It can be defined in the following way in the URDF

  .. code-block:: xml

    <joint name="right_finger_joint" type="prismatic">
      <axis xyz="0 1 0"/>
      <origin xyz="0.0 -0.48 1" rpy="0.0 0.0 0.0"/>
      <parent link="base"/>
      <child link="finger_right"/>
      <limit effort="1000.0" lower="0" upper="0.38" velocity="10"/>
    </joint>
    <joint name="left_finger_joint" type="prismatic">
      <mimic joint="right_finger_joint" multiplier="1" offset="0"/>
      <axis xyz="0 1 0"/>
      <origin xyz="0.0 0.48 1" rpy="0.0 0.0 3.1415926535"/>
      <parent link="base"/>
      <child link="finger_left"/>
      <limit effort="1000.0" lower="0" upper="0.38" velocity="10"/>
    </joint>

Mimic joints are an abstraction of the real world: They can be used to describe

* simple parallel kinematics with linear dependencies of the joint positions

> * abstract complex groups of actuated joints, where the positions of several joints are directly controlled by low-level control loops and move synchronously.

Author: I don't have one, I just can imagine that this could be a usecase ;) Or wheels of a multi-wheel diffbot, or fans of a drone with rotating at the same speed...

> * abstract complex groups of actuated joints, where the positions of several joints are directly controlled by low-level control loops and move synchronously.

Reviewer: Maybe leave some examples here, then it would be clearer. The belt and pulley system could also be a  example?

> * abstract complex groups of actuated joints, where the positions of several joints are directly controlled by low-level control loops and move synchronously.

Author: what do you think about the last update?

      <state_interface name="position"/>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>
  </ros2_control>

From the officially released packages, the following packages are already using this information:

* :ref:`mock_components (generic system) <mock_components_userdoc>`
* :ref:`gazebo_ros2_control <gazebo_ros2_control>`
* :ref:`gz_ros2_control <gz_ros2_control>`

As the URDF specifies only the kinematics, the mimic tag has to be independent of the hardware interface type used in ros2_control. This means that we interpret this info in the following way:

* **position = multiplier * other_joint_position + offset**
* **velocity = multiplier * other_joint_velocity**

If someone wants to deactivate the mimic joint behavior for whatever reason without changing the URDF, it can be done by setting the attribute ``mimic=false`` of the joint tag in the ``<ros2_control>`` section of the XML.

.. code-block:: xml

  <joint name="left_finger_joint" mimic="false">
    <state_interface name="position"/>
    <state_interface name="velocity"/>
    <state_interface name="effort"/>
  </joint>

Transmission Interface
#######################

> Mechanical transmissions transform effort/flow variables such that their product (power) remains constant. Effort variables for linear and rotational domains are force and torque; while the flow variables are respectively linear velocity and angular velocity.

Author: This was copied from here :D 
 https://github.com/ros-controls/ros2_control/blob/16fbde3d7ee9e6d7ce19c789887eb4558118f59b/transmission_interface/include/transmission_interface/transmission.hpp#L29-L38 
 
 Any suggestions? Then I'd change that in the docstring, too.

> Mechanical transmissions transform effort/flow variables such that their product (power) remains constant. Effort variables for linear and rotational domains are force and torque; while the flow variables are respectively linear velocity and angular velocity.

Reviewer: Well if it is like that in the original docs it's fine. 
At least for me the product he is talking about is not clear, I see that point of energy conservation, but within the transmission plugin you transform also position and velocity

In robotics it is customary to place transmissions between actuators and joints. This interface adheres to this naming to identify the input and output spaces of the transformation. The provided interfaces allow bidirectional mappings between actuator and joint spaces for effort, velocity and position. Position is not a power variable, but the mappings can be implemented using the velocity map plus an integration constant representing the offset between actuator and joint zeros.

> The ``transmission_interface`` provides a base class and some implementations for plugins, which can be integrated and loaded by custom hardware components. They are not automatically loaded by any hardware component or the gazebo plugins.

Reviewer: ```suggestion
The ``transmission_interface`` provides a base class and some implementations for plugins, which can be integrated and loaded by custom hardware components. They are not automatically loaded by any hardware component or the gazebo plugins, each hardware component is responsible for loading the appropriate transmission interface to map the actuator readings to joint readings.
```
